File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const n = nums.length;
    const res = new Array(n);

    let prefix = 1;
    for (let i = 0; i < n; i++) {
        res[i] = prefix;
        prefix = nums[i] * prefix;
    }

    let suffix = 1;
    for (let i = n - 1; i >= 0; i--) {
        res[i] = res[i] * suffix;
        suffix = nums[i] * suffix;
    }

    return res
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # product of array except self
        n = len(nums)
        res = [1] * n

        # calculate the prefix produc (all elements to the left)
        prefix = 1
        for i in range(n):
            res[i] = prefix
            prefix *= nums[i]

        # Multiply by the suffix produc (all elements to the right)
        suffix = 1
        for i in range(n -1, -1, -1):
            res[i] *= suffix # we multiply the prefix and postfix together
            suffix *= nums[i]

        return res